### PR TITLE
test: Multi-Publisher Plugin Unit Tests - Interim PR

### LINKED ISSUE #1419: Multi-Publisher Plugin Unit Tests
### Story Form

As a Block Node Engineer/Operator/User
I want to have unit tests for my classes
So that I can have confidence in existing logic and catch regressions

### Technical Notes

**DEPENDS ON**: #1370
- we have some unit tests already
- we must add additional unit tests to boost confidence

### PR DESCRIPTION
**Pull request by ata-nas**
## Reviewer Notes

* Added unit tests:
  * PublisherHandler:
    * test onNext receives subsequent requests if previous one expected a header but did not receive
    * tests for broken header supplied
    * tests for null header bytes supplied
    * test for sending acknowledgements
    * tests for closeBlock called
    * tests for handling block proof
  * StreamPublisherPlugin:
    * test for null items
    * test for empty items
    * test for unset oneOf
    * test for valid block (disabled, finish in a follow up)
* Updated Publisher Manager test fixture:
  * added implementation for handlePersisted
  * added implementation for closeBlock
  * added the ability to add a handler to send notifications to
* Extended Test Block Items Builder
  * added the ability to create blocks with broken headers
  * added the ability to create blocks with broken proofs
* Test Dependencies:
  * added jupiter params

**UPDATE**:
In order to unblock progress for simulator tasks, we will stop with tests here and make part 2: #1435
Note that some tests will be disabled because they require additional logic which will further complicate this and slow down simulator tasks even more.
Mainly, we need to tackle #1374 as it becomes even more problematic, we cannot test properly the plugin itself.

### CODE REVIEW

File: block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/PublisherHandlerTest.java [hunk 1]
```diff
@@ -240,6 +247,133 @@ void testPublishItemsNoRepliesHappyPathACCEPT() {
...
+         * We expect that when the {@link StreamPublisherManager} returns
+         * any action for the streamed block number and when the
+         * streamed items end with a valid block proof, the handler will send
+         * a signal to the manager to close the block.
+         */
+        @ParameterizedTest
+        @EnumSource(value = BlockAction.class)
+        @DisplayName(
+                "Test onNext() with valid request with a complete single block calls closeBlock on manager when batch ends with valid proof - happy path ACCEPT")
+        void testPublishItemsNullCloseBlockCallWhenRequestEndsWithBrokenProofHappyPathACCEPT(final BlockAction action) {
```
Comment (jsync-swirlds) [inline block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/PublisherHandlerTest.java:356]: (This is a nitpick and opinion, which can be discussed later if desired).

We have a description, the method name shouldn't try to include the entire description (it becomes useless because most tooling abbreviates it at this length).
Something much shorter and describing what _requirement_ is verified rather than the exact scenario(s) covered is best.
That makes it easier to combine scenarios too, a lot of these methods could be one parameterized method, if we name methods according to purpose rather than action.

File: block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/PublisherHandlerTest.java [hunk 2]
```diff
@@ -1096,6 +1230,174 @@ void testPublishReturnInvalidRequestNoMetricsUpdated() {
             assertThat(metrics.endStreamsReceived().get()).isEqualTo(0);
         }
 
+        /**
+         * This test aims to assert that the {@link PublisherHandler} correctly
+         * handles a received invalid request
```
Comment (jsync-swirlds) [inline block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/PublisherHandlerTest.java:1235]: (Continued nits about method naming, for future reference only).

This is a good example.
This method could be named `verifyHandlesInvalidRequest` except that's not actually what this validates (which suggests this summary needs change).
What the method appears to verify is "Handle streaming in the middle of a block by ignoring everything before a valid block header".
A good name would be `verifyCorrectHandlingWhenStreamsStartMidBlock` (still long, but much improved).  That could then run multiple scenarios to ensure that block streams that start mid-block (between header and proof) are handled correctly.

File: block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/fixtures/TestStreamPublisherManager.java [hunk 3]
```diff
@@ -41,14 +54,87 @@ public BlockAction getActionForBlock(
...
+        if (newLastPersistedBlock >= latestBlockNumber) {
+            publisherHandlers.forEach(h -> h.sendAcknowledgement(newLastPersistedBlock));
+        } else {
+            final String message = ("Illegal state for publisher manager test fixture. "
+                            + "`handlePersisted` is called when `latestBlockNumber` "
+                            + "is greater than the argument notification's end block number. "
+                            + "This is not allowed in fixtures, the latest block number must always be "
+                            + "set explicitly to a valid value before calling `handlePersisted` in order "
+                            + "to mitigate false positives. \n"
+                            + "latestBlockNumber: %d, notification end block number: %d")
```
Comment (jsync-swirlds) [inline block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/fixtures/TestStreamPublisherManager.java:98]: Instead of a bunch of concatenation, perhaps use a text block here?
Also, this can perhaps be a constant just before the method for more clarity within the method.

File: block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/StreamPublisherPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        private final Function<PublishStreamResponse, ResponseOneOfType> responseKindExtractor =
+                response -> response.response().kind();
+        private final Function<PublishStreamResponse, Code> endStreamResponseCodeExtractor =
+                response -> Objects.requireNonNull(response.endStream()).status();
+
+        /**
+         * Constructor for the plugin tests.
+         */
+        PluginTest() {
+            super(new BlockingSerialExecutor(new LinkedBlockingQueue<>()));
```
Comment (jsync-swirlds) [inline block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/StreamPublisherPluginTest.java:70]: Definitely should not be using a serial blocking executor here.

This can be addressed in a future PR, however.

**Review by jsync-swirlds — APPROVED** [review]
Some concerns that should be addressed in the next PR, but nothing that blocks merging this interim PR.

File: block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java [hunk 5]
```diff
@@ -41,72 +39,47 @@
...
-    /**
-     * Start the test fixture.<br/>
-     * This overload uses the default configuration and the default
-     * test thread pool manager.
-     *
-     * @param plugin the plugin to be tested
-     * @param historicalBlockFacility the historical block facility to be used
-     */
-    public void start(P plugin, HistoricalBlockFacility historicalBlockFacility) {
-        start(plugin, historicalBlockFacility, testThreadPoolManager, null);
```
[the remark this answers is not in the document]
Comment (ata-nas) [reply block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java:68]: I was thinking about this for a very long time.
This is very deliberate. In the plugin fixtures we utilize a fixture for the thread pool manager. We need this fixture so we can have control over the supplied thread pool. I do not see that ever changing. The thread pool yes, but not the manager fixture.
I decided to remove this because I noticed quite a few issues, for example the override changed the plugin-under-test's thread pool manager, but not the fixture's one. This lead to multiple issues for me. Moreover, we were using a thread pool type by default until now, but when something is not explicitly set (like forcing the user to call a constructor) it is often overlooked. I wanted to make thread pool type explicit for tests as we do need to take heed in what pool we are using (as mentioned in other places).
I do see the argument where we will want to change the thread pool manager fixture itself, but I think it's place is in a constructor, a single place that you cannot miss. But I do argue that if we want to change the thread pool manager fixture implementation (which I very highly doubt), we can very easily do it through the constructor, the refactor would be minimal.

File: block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java [hunk 6]
```diff
@@ -41,72 +39,47 @@
...
-     * @param testThreadManager the thread pool manager to be used
      * @param configOverrides a map of configuration overrides to be applied to loaded configuration
      */
     public void start(
-            P plugin,
-            HistoricalBlockFacility historicalBlockFacility,
-            ThreadPoolManager testThreadManager,
-            Map<String, String> configOverrides) {
+            @NonNull final P plugin,
+            @NonNull final HistoricalBlockFacility historicalBlockFacility,
```
[the remark this answers is not in the document]
Comment (ata-nas) [reply block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java:81]: I see value here, these will be called and will produce an NPE if the args are null, simply no way around it. The value I see here is that the user will see this and know they need to supply a non-null value.

File: block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginTest.java [hunk 7]
```diff
@@ -44,6 +46,7 @@ class BackfillPluginTest extends PluginTestBase<BackfillPlugin> {
             "1000");
 
     public BackfillPluginTest() {
+        super(new BlockingSerialExecutor(new LinkedBlockingQueue<>()));
```
[the remark this answers is not in the document]
Comment (ata-nas) [reply block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginTest.java:49]: I'd say in places where we use pools to release core application threads (like messaging) faster is absolutely ok to use them. The benefit of this pool is that we can easily debug issues during tests and also we block until tasks execute which ensures our tests will not be flaky, we will be asserting at the correct time.

File: block-node/s3-archive/src/test/java/org/hiero/block/node/archive/s3/S3ArchivePluginTest.java [hunk 8]
```diff
@@ -56,6 +57,7 @@ class S3ArchivePluginTest extends PluginTestBase<S3ArchivePlugin> {
 
     @SuppressWarnings("resource")
     public S3ArchivePluginTest() throws Exception {
+        super(new BlockingSerialExecutor(new LinkedBlockingQueue<>()));
```
[the remark this answers is not in the document]
Comment (ata-nas) [reply block-node/s3-archive/src/test/java/org/hiero/block/node/archive/s3/S3ArchivePluginTest.java:60]: The runtime implementation uses a single threaded pool. The way this plugin utilizes the pool is that it simply queues up tasks to be executed in order of submission. Each task decides if it will execute something. I think this would actually be a perfect candidate for the blocking executor, the only purpose of the thread pool at runtime is to release the messaging thread ASAP, as uploads could take a very long time.

Mind you, until now, we used the blocking executor by default in tests. Hence after this refactor I used that by default, except in cases where the pool was overwritten in tests.

File: block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/PublisherHandlerTest.java (hunk 2 above)
Comment (ata-nas) [reply block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/PublisherHandlerTest.java:1235]: Thanks for the suggestion, I will attempt to improve these in the follow up. 🙌

File: block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/StreamPublisherPluginTest.java (hunk 4 above)
Comment (ata-nas) [reply block-node/stream-publisher/src/test/java/org/hiero/block/node/stream/publisher/StreamPublisherPluginTest.java:70]: I've already changed this in the follow up. It is not even possible to use this because the config we use for iterations of the task holds for a very long time, I cannot change it below 100k as well.

**Review by AlfredoG87 — APPROVED** [review]

File: block-node/s3-archive/src/test/java/org/hiero/block/node/archive/s3/S3ArchivePluginTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply block-node/s3-archive/src/test/java/org/hiero/block/node/archive/s3/S3ArchivePluginTest.java:60]: Using the single threaded pool is not ideal, and I suspect we'll need to change that soonish.
What I hope we can do is avoid testing in single thread mode so we have the option to use multiple threads in pools.

File: block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java:81]: Annotations are not _called_, they have no effect at runtime.

Also, in test fixtures it is entirely reasonable to set null for some of these in order to set up scenarios at the edge.

File: block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ata-nas) [reply block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java:68]: We are not, we can freely choose the thread pool setup.

File: block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginTest.java [hunk 9]
```diff
@@ -24,7 +26,7 @@
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
 
-class BackfillPluginTest extends PluginTestBase<BackfillPlugin> {
+class BackfillPluginTest extends PluginTestBase<BackfillPlugin, BlockingSerialExecutor> {
```
[the remark this answers is not in the document]
Comment (jsync-swirlds) [reply block-node/backfill/src/test/java/org/hiero/block/node/backfill/BackfillPluginTest.java:29]: I actually prefer to have a simpler fixture (because test code is 100% overhead with no purpose beyond testing) and have the individual tests track things a bit more.

Perhaps something to discuss in channel and get a read for the whole team?
